src/profile-logic/profile-data.js

```diff
@@ -2911,3 +2915,444 @@ export function hasThreadKeys(
+
+  // Samples
+  for (let i = 0; i < samples.length; i++) {
+    const stack = samples.stack[i];
+    if (stack !== null) {
+      samplingSelfStacks.add(stack);
+    }
+  }
+
+  // Markers
+  for (let i = 0; i < markers.length; i++) {
+    const data = markers.data[i];
+    if (data && data.cause) {
+      const stack = data.cause.stack;
+      if (stack !== null) {
+        syncBacktraceSelfStacks.add(stack);
+      }
+    }
+  }
+
+  // JS allocations
+  if (jsAllocations !== undefined) {
+    for (let i = 0; i < jsAllocations.length; i++) {
+      const stack = jsAllocations.stack[i];
+      if (stack !== null) {
+        syncBacktraceSelfStacks.add(stack);
+      }
+    }
+  }
+
+  // Native allocations
+  if (nativeAllocations !== undefined) {
+    for (let i = 0; i < nativeAllocations.length; i++) {
+      const stack = nativeAllocations.stack[i];
+      if (stack !== null) {
+        syncBacktraceSelfStacks.add(stack);
+      }
+    }
+  }
+
+  return { samplingSelfStacks, syncBacktraceSelfStacks };
+}
+
+/**
+ * Create a new thread with all stack references translated via the given
+ * maps. The maps map IndexIntoOldStackTable -> IndexIntoNewStackTable.
+ * Only a subset of entries are read from the map: The same stacks that
+ * gatherStackReferences found in the thread. All other entries are ignored
+ * and do not need to be present.
+ * With the exception of the caller which does the initial return address
+ * nudging, most callers will want to pass the same map to both map arguments.
+ */
+export function replaceStackReferences(
+  thread: Thread,
+  mapForSamplingSelfStacks: Map<IndexIntoStackTable, IndexIntoStackTable>,
+  mapForBacktraceSelfStacks: Map<IndexIntoStackTable, IndexIntoStackTable>
+): Thread {
+  const {
+    samples: oldSamples,
+    markers: oldMarkers,
+    jsAllocations: oldJsAllocations,
+    nativeAllocations: oldNativeAllocations,
+  } = thread;
+
+  // Samples
+  const samples = {
+    ...oldSamples,
+    stack: oldSamples.stack.map((oldStackIndex) => {
+      if (oldStackIndex === null) {
+        return null;
+      }
+      const newStack = mapForSamplingSelfStacks.get(oldStackIndex);
+      if (newStack === undefined) {
+        throw new Error(
+          `Missing mapForSamplingSelfStacks entry for stack ${oldStackIndex}`
+        );
+      }
+      return newStack;
+    }),
+  };
+
+  // Markers
+  function replaceStackReferenceInPayload(
+    oldData: MarkerPayload
+  ): MarkerPayload {
+    if (oldData && 'cause' in oldData && oldData.cause) {
+      // Replace the cause with the right stack index.
+      // Use (...: any) because our current version of Flow has trouble with
+      // the object spread operator.
+      const oldStackIndex = oldData.cause.stack;
+      const newStack = mapForBacktraceSelfStacks.get(oldStackIndex);
+      if (newStack === undefined) {
+        throw new Error(
+          `Missing mapForBacktraceSelfStacks entry for stack ${oldStackIndex}`
+        );
+      }
```

Comment:
Is it worth reusing `mapBacktraceSelfStack`?

Not for saving lines, more because this is surprising we don't use it here, and then I got to wonder if there's a difference.

Reply:
Good idea. No reason, I just hadn't thought of it.